### ISSUE TITLE
net/haproxy: the tcp-check syntax used to look for a custom string is broken

### ISSUE THREAD

**Issue by ladar** [opnsense/plugins#2653]
**Describe the bug**

The config files being generated by the Haproxy plugin are invalid, when a custom TCP check has been setup to match a given string.

It appears the template is putting the "send" and "expect" options on the same line, which might have worked in the past but doesn't any more. I upgraded one of my firewalls in July and that broke the existing config. I had to disable the string checks as a result.

Currently the plugin is producing this output:

```
backend DMTP
    option log-health-checks
    # health check: DMTP
    option tcp-check
    tcp-check connect
    tcp-check send QUIT expect string "220 lavabit.com DSMTP Magma"
    mode tcp
```

Note the  line **`tcp-check send QUIT expect string "220 lavabit.com DSMTP Magma"`** ... that line is invalid and produces the following error:

```
redacted@edge:~ # haproxy -c -f /usr/local/etc/haproxy.conf.staging
[NOTICE] 319/135742 (15247) : haproxy version is 2.2.17-dd94a25
[ALERT] 319/135742 (15247) : parsing [/usr/local/etc/haproxy.conf.staging:164] : 'tcp-check send' : expects 'comment' but got 'expect' as argument..
[ALERT] 319/135742 (15247) : Error(s) found in configuration file : /usr/local/etc/haproxy.conf.staging
[ALERT] 319/135742 (15247) : Fatal errors found in configuration.
```

The syntax for this backend check should be:

```
    tcp-check send QUIT 
    tcp-check expect string "220 lavabit.com DSMTP Magma"
```

Put simply, the expect string needs to be on a separate line, and prefaced with `tcp-check`. All of the examples I found online placed the `tcp-check expect` line immediately after the `tcp-check send` line. Using that layout, you can include multiple send/expect checks in sequence. 
 
Also note that because the string I'm matching against contains whitespace, I had to manually add the quotes. Without quotes, or backslashes preceding each space, the parser assumes the matching string is only 220.

My suggestion would be to simply quote every arbitrary user input string. At least until a more robust engine is created to properly escape what needs to be escaped.

Here is a screenshot of what this backend looks like in a browser:

![Screenshot from 2021-11-16 13-55-58](https://user-images.githubusercontent.com/1434008/142057770-ca31a254-124a-4a81-90b8-f6f336635fc5.png)

**Important notices**
Before you add a new report, we ask you kindly to acknowledge the following:

[X] I have read the contributing guide lines at https://github.com/opnsense/plugins/blob/master/CONTRIBUTING.md

[X] I have searched the existing issues and I'm convinced that mine is new.

[X The title contains the plugin to which this issue belongs

**Environment**

I'm currently using:

OPNsense 21.7.5-amd64
FreeBSD 12.1-RELEASE-p21-HBSD
LibreSSL 3.3.5

And the haproxy packages are:

os-haproxy 3.7
haproxy 2.2.17

I believe these are the latest versions available. As for when my configuration got flubbed, that happened earlier this year, when I made the mistake of migrating from 20.7.8_4 to 21.1.8. That was in July. The haproxy package was updated at that time as well, and I can only presume that the newer haproxy runtime applies a more stringent Backus normal format. 

I've cherry picked what I think are the most relevant lines from my system log:

```
2021-07-20T21:21:16	pkg-static[10211]	haproxy-2.2.14 installed
2021-07-20T21:20:54	pkg-static[10211]	opnsense-update upgraded: 21.1 -> 21.1.8
2021-07-20T20:54:25	pkg-static[20052]	opnsense upgraded: 20.7.8_4 -> 21.1
2021-07-20T19:46:57	pkg-static[36278]	opnsense-update upgraded: 20.7.6 -> 20.7.8	
2020-12-09T23:33:12	pkg-static[28385]	opnsense upgraded: 20.7.2 -> 20.7.6
```

Note that before the upgrade in July, when I went from 20.7.2 to 21.1.8, my custom TCP checks were working properly (I tested them during the initial setup). I had several checks which sent strings containing whitespace (for example the IMAP checks sent `A01 LOGOUT` and several more matching strings with spaces/asterisks, etc. And they all worked just fine without needing to enclose them in quotes via the webgui. I can't speack to whether an older plugin version wrapped my inputs in quotes, or haproxy didn't care. All I know is they worked. And that was enough for me.

I updated my standby firewall system this week, and was hoping this issue would be fixed by now, but it appears to still be a problem. I believe the syntax can fixed in this file `[/usr/local/opnsense/service/templates/OPNsense/HAProxy/haproxy.conf](https://github.com/opnsense/plugins/blob/master/net/haproxy/src/opnsense/service/templates/OPNsense/HAProxy/haproxy.conf)` but don't have the time to fix it right now.

**Comment by ladar** [comment]
@fraenki I just noticed your post regarding the upgrade haproxy upgrade. I like new features. But is there any chance you could fix, and possibly improve the interface for creating up backend checks? I'd be happy with ability to chain multiple `tcp-check send` and `tcp-check expect` lines together. Something like:

```
tcp-check expect string "* OK lavabit.com Magma IMAP server"
tcp-check send "A01 CAPABILITY\r\n"
tcp-check expect string "* CAPABILITY IMAP4 IMAP4rev1 LITERAL+ ID\r\nA01 OK Completed."
tcp-check send "A02 LOGOUT\r\n"
tcp-check expect string "* BYE Goodbye.\r\nA02 OK Completed."
```
Note the first `expect` should check the banner. While the others ensure the server is running properly and responding.

Pretty please?

**Comment by fraenki** [comment]
I'll look into the reported bug (which may take some time), but...

> ...improve the interface for creating up backend checks? I'd be happy with ability to chain multiple tcp-check send and tcp-check expect lines together.

...I've already said it [elsewhere](https://github.com/opnsense/plugins/pull/2560#issuecomment-938144942): we need more UI contributors. There's plenty of room for UI improvements in the HAProxy plugin. However, all this UI and Javascript magic exceed my skills, so I simply cannot contribute these improvements. So until someone steps in to work on UI improvements, the UI will remain as it is. :(

Don't get me wrong, I like the idea of chaining multiple checks and plenty of other great ideas were already suggested, but someone needs to volunteer and work on it to make it happen :)

**Comment by ladar** [comment]
> ...I've already said it elsewhere: we need more UI contributors. There's plenty of room for UI improvements in the HAProxy plugin. However, all this UI and Javascript magic exceed my skills, so I simply cannot contribute these improvements. So until someone steps in to work on UI improvements, the UI will remain as it is. :(

Yeah, sadly, my web development skills are exceedingly rusty. Which is why, even though I roughly know where the "tcp-check send:" lines are being generated, I don't have the skills to make changes without spending some time looking over the existing code to understand how it works.

That said, I'd settle for a raw input box. I already know what I want add/change, but I don't know how to add those lines to the backend config and have them persist. Something like the `Option pass-through` text area on the virtual server dialog. I was able to add `source ipv4@ usesrc clientip` to it, and along with a lot of `ipfw` magic, my mail servers can see the actual IP of incoming connections, which is critical for email. One of these days, I'll type up what I learned as a feature request/issue, but that will take a fair bit of typing. 

Anyways, could we perhps just add another `Option pass-through` to the backend config screen? As long as it supports multiple lines (ie text area not text box), it will provide a nice fall back for advanced users to workaround issues like the broken config syntax in the future.

**Comment by fraenki** [comment]
> Anyways, could we perhps just add another Option pass-through to the backend config screen?

The use of the existing "pass-through" options is strongly discouraged. Such configs are pretty likely to break at one point in the future during an upgrade of OPNsense or HAProxy. I understand that it may solve an immediate need, but it causes trouble long term. Hence I will not add it, it needs to be properly implemented.

**Comment by ladar** [comment]
@fraenki I'm all for a proper implementation via the GUI. And I don't want to sound crass, but when it comes down to it, if the GUI doesn't provide critical functionality, then my choices are abandon the plugin altogether and use/manage haproxy at the OS level, or switch back to using pfsense. I really don't like either option. Which is why I've spent hours getting my IP pass through to work by reverse engineering how the pfsense plugin handled it, and then adding the requisite haproxy config options to the pass-through, and creating my on `ipfw` configs. 

We use OPNsense for route, to handle firewall/nating, and to handle load balancing. And the ability to pass through a client's IP is critical absolutely critical for load balancing email servers. Likewise the ability to differentiate between a server that is accepting TCP connections, versus one that is actually working is almost as critical.

**Comment by fraenki** [comment]
> And I don't want to sound crass, but when it comes down to it, if the GUI doesn't provide critical functionality, then my choices are abandon the plugin altogether

Sure, that's your choice. I'm not selling a product here or anything, I'm just volunteering and I have to decide where I'll spend my time on. So that's my choice. :)

**Comment by fraenki** [comment]
Thanks for reporting this issue. I've fixed the tcp-check syntax in https://github.com/opnsense/plugins/pull/2744/commits/973ad0987371b0a5127efeb4a7b7012ba22237a4. This will be available in the upcoming os-haproxy 3.9.
